src/shogun/mathematics/linalg/backend/eigen/EigenSolvers.cpp

```diff
@@ -0,0 +1,166 @@
+
+	/*
+	 * checking for success
+	 *
+	 * 0: Eigen::Success. Decomposition was successful
+	 * 1: Eigen::NumericalIssue. The input contains INF or NaN values or
+	 * overflow occured
+	 */
+	REQUIRE(
+	    solver.info() != Eigen::NumericalIssue,
+	    "The input contains INF or NaN values or overflow occured.\n");
+
+	eigenvalues_eig = solver.eigenvalues().real();
+	eigenvectors_eig = solver.eigenvectors().real();
+}
+
+void LinalgBackendEigen::eigen_solver_impl(
+    const SGMatrix<complex128_t>& A, SGVector<complex128_t>& eigenvalues,
+    SGMatrix<complex128_t>& eigenvectors) const
+{
+	typename SGMatrix<complex128_t>::EigenMatrixXtMap A_eig = A;
+	typename SGMatrix<complex128_t>::EigenMatrixXtMap eigenvectors_eig =
+	    eigenvectors;
+	typename SGVector<complex128_t>::EigenVectorXtMap eigenvalues_eig =
+	    eigenvalues;
+
+	Eigen::ComplexEigenSolver<typename SGMatrix<complex128_t>::EigenMatrixXt>
+	    solver(A_eig);
+
+	REQUIRE(
+	    solver.info() != Eigen::NumericalIssue,
+	    "The input contains INF or NaN values or overflow occured.\n");
+
+	eigenvalues_eig = solver.eigenvalues();
+	eigenvectors_eig = solver.eigenvectors();
+}
+
+template <typename T>
+void LinalgBackendEigen::eigen_solver_symmetric_impl(
+    const SGMatrix<T>& A, SGVector<T>& eigenvalues, SGMatrix<T>& eigenvectors,
+    index_t k) const
+{
+	typename SGMatrix<T>::EigenMatrixXtMap A_eig = A;
+	typename SGMatrix<T>::EigenMatrixXtMap eigenvectors_eig = eigenvectors;
+	typename SGVector<T>::EigenVectorXtMap eigenvalues_eig = eigenvalues;
+
+	Eigen::SelfAdjointEigenSolver<typename SGMatrix<T>::EigenMatrixXt> solver(
+	    A_eig);
+
+	/*
+	 * checking for success
+	 *
+	 * 0: Eigen::Success. Eigenvalues computation was successful
+	 * 2: Eigen::NoConvergence. Iterative procedure did not converge.
+	 */
+	REQUIRE(
+	    solver.info() != Eigen::NoConvergence,
+	    "Iterative procedure did not converge!\n");
+
+	eigenvalues_eig = solver.eigenvalues().tail(k).template cast<T>();
+	eigenvectors_eig = solver.eigenvectors().rightCols(k).template cast<T>();
+}
+
+#ifdef HAVE_LAPACK
+template <>
+void LinalgBackendEigen::eigen_solver_symmetric_impl<float64_t>(
+    const SGMatrix<float64_t>& A, SGVector<float64_t>& eigenvalues,
+    SGMatrix<float64_t>& eigenvectors, index_t k) const
+{
+	int32_t status = 0;
+	int32_t n = A.num_rows;
+
+	// dsyevr requires a vector of length n even if you want just k eigenvalues
+	SGVector<float64_t>::EigenVectorXt ev_eig(n);
+	wrap_dsyevr(
+	    'V', 'U', n, A.matrix, n, n - k + 1, n, ev_eig.data(),
+	    eigenvectors.matrix, &status);
+
+	typename SGVector<float64_t>::EigenVectorXtMap eigenvalues_eig =
+	    eigenvalues;
+	eigenvalues_eig = ev_eig.head(k);
+
+	/*
+	 * checking for success
+	 *
+	 * status == 0: successful exit
+	 * status < 0: the i-th argument had an illegal value
+	 * status > 0: internal error
+	 */
+	REQUIRE(!(status < 0), "The %d-th argument han an illegal value.", -status)
```

Comment:
Very bad error messages (not useful for user). When printing errors, always assume the user of an algorithm gets them and has no idea what is wrong. So "The %d-th argument han an illegal value." is not optimal here